intfuncs: short-circuit gcd

tkluck opened the pull request:
When one argument to gcd is equal to 1, the result is fixed and there is no need to compute the rest.

This makes an enormous difference on random vectors of integers:

    julia> A = BigInt.(rand(Int, 2<<22));

    julia> gcd(A); @time gcd(A)
      0.404063 seconds (16.78 M allocations: 448.000 MiB)
    1

    julia> _gcd(A) = (res=zero(eltype(A)); for a in A; res=gcd(res,a); if res == 1 return res; end; end; return res)
    _gcd (generic function with 1 method)

    julia> _gcd(A); @time _gcd(A)
      0.000028 seconds (38 allocations: 1.086 KiB)
    1

but that's probably not a fair representation of real-world impact, as no such improvement is to be expected unless we reach 1 very quickly. A random vector will do that in very few iterations.

tkluck committed:
intfuncs: short-circuit gcd

When one argument to gcd is equal to 1, the result is fixed
and there is no need to compute the rest.

This makes an enormous difference on random vectors of integers:

    julia> A = BigInt.(rand(Int, 2<<22));

    julia> gcd(A); @time gcd(A)
      0.404063 seconds (16.78 M allocations: 448.000 MiB)
    1

    julia> _gcd(A) = (res=zero(eltype(A)); for a in A; res=gcd(res,a); if res == 1 return res; end; end; return res)
    _gcd (generic function with 1 method)

    julia> _gcd(A); @time _gcd(A)
      0.000028 seconds (38 allocations: 1.086 KiB)
    1

but that's probably not a fair representation of real-world impact, as
no such improvement is to be expected unless we reach 1 very quickly. A
random vector will do that in very few iterations.

tkluck committed:
Add testset for array version of gcd

I'm adding this because I added a new (shortcircuiting) codepath
for evaluating this. However, it is also generally useful because
it adds the case of empty collection (`gcd(Int[]) === 0`) as a testable
case.

tkluck commented:
This also enables `gcd(Int[])`, which should give `0` instead of

    ArgumentError: reducing over an empty collection is not allowed

I've just added test cases to test this new behaviour.

Similarly, `lcm(Int[])` should probably return `1`. I'll push that as well if you want; just let me know whether the pull request looks sane to you before I invest more time in it.

StefanKarpinski reviewed (approved):
Makes sense. For the many-valued case, it's a straightforward optimization; for the zero-valued case, it seems like the right definition. The usual worry is giving the correct zero type for the empty case.

tkluck committed:
intfuncs: add empty case for lcm(Integer[])

This should be `1`, ensuring that

    lcm(union(A,B)) == lcm(lcm(A),lcm(B))

Also adding test cases.

tkluck commented:
Thanks @StefanKarpinski . I took your comment as a cue to write+push the `lcm` part as well.

JeffBezanson commented on test/intfuncs.jl:
```diff
@@ -35,6 +35,31 @@
+        @test gcd(T[0, 15]) === T(15)
+        @test gcd(T[3,-15]) === T(3)
+        @test gcd(T[-3,-15]) === T(3)
+        @test gcd(T(0, 0)) === T(0)
```
Should be `T[0, 0]`.

tkluck replied:
Thanks to you (and to Travis :) ) for catching that! Just pushed a fix.

tkluck committed:
Fix mistake in new array gcd tests

Thanks to @JeffBezanson for catching it!

tkluck commented:
Looks like the failing CI are timeouts and manual cancellations. What's the best way to try again and get this merged?

StefanKarpinski commented:
I've restarted the AppVeyor builds. Travis Linux tests passed and FreeBSD passed so if one of the Windows runs passes too, we can merge. (Be sure to squash!)

tkluck commented:
One of the appveyor builds worked, and the other failed with an ostensibly unrelated error:
```
Error in testset libgit2:
Test Failed at C:\projects\julia\julia-9af421150a\share\julia\test\libgit2.jl:1592
  Expression: username == "foo"
   Evaluated: "bar" == "foo"
ERROR: LoadError: Test run finished with errors
in expression starting at C:\projects\julia\julia-9af421150a\share\julia\test\runtests.jl:52
```
Squash & merge?
